### PR TITLE
feat(tasks,emails): add search subcommands with filter/sort support

### LINKED ISSUE #53: feat: add search subcommands for tasks and emails with filter/sort support
## Problem

`hspt tasks list` and `hspt emails list` return paginated, unfiltered results with no way to query by status, owner, direction, or date. Finding open tasks for a specific owner requires paginating through the entire task history — potentially thousands of records — and filtering client-side.

```bash
# What you'd want (doesn't exist upstream):
hspt tasks search --filter "hs_task_status=NOT_STARTED" --filter "hubspot_owner_id=77999105"
hspt emails search --filter "hs_email_direction=EMAIL" --sort "hs_timestamp:desc"

# What you have to do instead:
hspt tasks list --limit 100  # oldest-first, no status filter, repeat with --after...
```

## Precedent

`contacts search` and `deals search` already use the HubSpot CRM Search API. The filter/sort infrastructure in `internal/cmd/shared/` already handles all the heavy lifting:

- `ParseFilters()` — parses `--filter` flags into `SearchFilterGroup` structs
- `ParseSort()` — parses `--sort` flags into `SearchSort` structs
- Auto-converts ISO dates to Unix milliseconds
- Supports shorthand (`prop=val`, `prop>=val`) and explicit operators (`prop:BETWEEN:low:high`)

## Proposed API

```bash
# Tasks: open tasks for a specific owner
hspt tasks search --filter "hs_task_status=NOT_STARTED" --filter "hubspot_owner_id=77999105" --sort "hs_timestamp:asc" --limit 50 -o json

# Tasks: overdue tasks
hspt tasks search --filter "hs_task_status=NOT_STARTED" --filter "hs_timestamp<=2026-03-17" --sort "hs_timestamp:asc"

# Emails: outbound emails
hspt emails search --filter "hs_email_direction=EMAIL" --sort "hs_timestamp:desc" --limit 20

# Emails: by subject
hspt emails search --filter "hs_email_subject:CONTAINS_TOKEN:Dev Academy" --limit 10
```

Both commands accept `--filter`, `--sort`, `--limit`, `--after`, and `--properties` flags — identical to `contacts search` and `deals search`.

## Implementation notes

Each new subcommand follows the `contacts search` pattern (`internal/cmd/contacts/contacts.go:newSearchCmd`):

1. Add `newSearchCmd(opts)` to the command's `Register()` function
2. Accept the standard search flags
3. Call `shared.ParseFilters()` and `shared.ParseSort()`
4. Call `client.SearchObjects(api.ObjectTypeTasks, req)` / `client.SearchObjects(api.ObjectTypeEmails, req)`
5. Render via `v.Render()`

The implementation is primarily wiring — no new infrastructure needed.

### PR DESCRIPTION
**Pull request by piekstra**
Closes #53

## Summary

Adds `hspt tasks search` and `hspt emails search`, backed by the HubSpot CRM Search API, with generic filtering and sorting — so you can find open tasks for an owner, overdue tasks, outbound emails, etc. without paginating the entire history and filtering client-side.

Both subcommands accept the same flags as the issue requested:

- `--filter` (repeatable)
- `--sort` (repeatable)
- `--limit`, `--after`, `--properties`

### Filter syntax

Shorthand comparisons:

| Form | Operator |
|------|----------|
| `prop=value` | `EQ` |
| `prop!=value` | `NEQ` |
| `prop>=value` | `GTE` |
| `prop<=value` | `LTE` |
| `prop>value` | `GT` |
| `prop<value` | `LT` |

Explicit operators (uppercase, HubSpot convention):

- `prop:OPERATOR:value` — e.g. `hs_email_subject:CONTAINS_TOKEN:Dev Academy`
- `prop:BETWEEN:low:high`
- `prop:IN:a,b,c` / `prop:NOT_IN:a,b,c`
- `prop:HAS_PROPERTY` / `prop:NOT_HAS_PROPERTY`

ISO-8601 dates supplied for known date properties (e.g. `hs_timestamp`) are auto-converted to Unix milliseconds, so `hs_timestamp<=2026-03-17` works as expected.

### Sort syntax

`prop`, `prop:asc`, `prop:desc` (case-insensitive; defaults to ascending).

### Examples

```bash
# Open tasks for a specific owner, oldest first
hspt tasks search --filter "hs_task_status=NOT_STARTED" --filter "hubspot_owner_id=77999105" --sort "hs_timestamp:asc"

# Overdue tasks
hspt tasks search --filter "hs_task_status=NOT_STARTED" --filter "hs_timestamp<=2026-03-17" --sort "hs_timestamp:asc"

# Outbound emails, newest first
hspt emails search --filter "hs_email_direction=EMAIL" --sort "hs_timestamp:desc" --limit 20

# Emails by subject
hspt emails search --filter "hs_email_subject:CONTAINS_TOKEN:Dev Academy" --limit 10
```

### Implementation notes

The issue references `shared.ParseFilters()` / `shared.ParseSort()` as already existing, but they did not exist in this repo (only `shared/format.go` was present, and the existing `contacts`/`deals`/`tickets` search commands use hardcoded per-field flags). This PR introduces those generic helpers in `internal/cmd/shared/search.go` and wires the two new subcommands to them.

- `api.SearchFilter` gains optional `highValue` (BETWEEN) and `values` (IN/NOT_IN) fields — additive and backward compatible; existing search commands are unaffected.
- New subcommands follow the existing `newSearchCmd` registration pattern and the table/JSON/plain rendering via `v.Render()`.

## Files changed

- `internal/cmd/shared/search.go` (new) — `ParseFilters` / `ParseSort`
- `internal/cmd/shared/search_test.go` (new) — parser unit tests
- `internal/cmd/tasks/tasks.go` — register + implement `search`
- `internal/cmd/emails/emails.go` — register + implement `search`
- `api/crm.go` — `SearchFilter.HighValue` / `.Values`
- `api/crm_test.go` — request-serialization tests for tasks/emails search (incl. BETWEEN/IN)
- `README.md` — document the new subcommands

## Testing

All commands run from the repo root.

```
$ go test -race ./...
ok  github.com/open-cli-collective/hubspot-cli/api
ok  github.com/open-cli-collective/hubspot-cli/internal/cmd/configcmd
ok  github.com/open-cli-collective/hubspot-cli/internal/cmd/shared
(all other packages: no test files)
# PASS

$ golangci-lint run ./...
0 issues.

$ make build
# succeeds -> bin/hspt
```

Manual smoke tests against `bin/hspt`:

- `hspt tasks search --help` / `hspt emails search --help` show the new flags and examples.
- Validation errors surface cleanly, e.g.:
  - `--sort "hs_timestamp:sideways"` -> `invalid sort direction "sideways" ... (use asc or desc)`
  - `--filter "hs_task_status"` -> `invalid filter "hs_task_status": expected prop=value, ...`

(Live API calls were not exercised — no HubSpot token in this environment — but request construction is covered by the `api` package tests.)

### CODE REVIEW

File: api/crm_test.go [hunk 1]
```diff
@@ -428,6 +429,67 @@ func TestClient_SearchObjects(t *testing.T) {
...
+			Properties: []string{"hs_task_subject"},
+			Limit:      50,
+			After:      "cursor123",
+		})
+
+		require.NoError(t, err)
+		assert.Len(t, result.Results, 1)
+		assert.Equal(t, float64(50), body["limit"])
+		assert.Equal(t, "cursor123", body["after"])
+		assert.Contains(t, body, "filterGroups")
```
Comment (piekstra-dev) [inline api/crm_test.go]: The "search tasks sends filter, sort, and pagination" sub-test only asserts that `filterGroups` and `sorts` keys are present in the request body — not their content. A serialisation bug that emitted `"filterGroups": [{}]` or omitted the filter operator entirely would still pass. This is the primary test path for a plain EQ filter round-trip through `SearchRequest` → JSON, and its correctness is now particularly load-bearing given the struct was just extended with `HighValue` and `Values`.

Add at least one content assertion on the filter:
```go
groups := body["filterGroups"].([]interface{})
f := groups[0].(map[string]interface{})["filters"].([]interface{})[0].(map[string]interface{})
assert.Equal(t, "hs_task_status", f["propertyName"])
assert.Equal(t, "EQ", f["operator"])
assert.Equal(t, "NOT_STARTED", f["value"])
```
The second sub-test already does this depth of inspection for BETWEEN/IN, so the pattern is established.

*Reply inline to this comment.*

File: internal/cmd/tasks/tasks.go [hunk 2]
```diff
@@ -361,6 +363,103 @@ func newDeleteCmd(opts *root.Options) *cobra.Command {
 	return cmd
```
Comment (piekstra-dev) [inline internal/cmd/tasks/tasks.go:363]: The `newSearchCmd` implementations in `tasks.go` and `emails.go` are nearly identical — they differ only in object type, table headers, and example text. The same PR that introduced `internal/cmd/shared` to hold reusable filter/sort logic did not extract the shared command scaffold, so any future object type (calls, contacts, etc.) would require a third copy.

Suggested fix: add a constructor in `internal/cmd/shared` (e.g. `NewSearchCmd(opts, objectType, headers, defaultProperties, examples)`) and delegate from each object-specific `newSearchCmd`. This keeps the reuse boundary consistent with what was already established by `ParseFilters`/`ParseSort`.

*Reply inline to this comment.*

**Comment by piekstra-dev** [comment]
## Automated PR Review

**Reviewed commit:** `d9d1e00cfba5`
**Profile:** `reviewer` - **Posting as:** `piekstra-dev`

### Summary

| Reviewer | Findings |
|----------|----------|
| documentation:docs | 0 |
| go:implementation-tests | 1 |
| policies:conventions | 1 |

<details>
<summary><strong>go:implementation-tests (1 finding)</strong></summary>

### Minor - `api/crm_test.go:460`

> The "search tasks sends filter, sort, and pagination" sub-test only asserts that `filterGroups` and `sorts` keys are present in the request body — not their content. A serialisation bug that emitted `"filterGroups": [{}]` or omitted the filter operator entirely would still pass. This is the primary test path for a plain EQ filter round-trip through `SearchRequest` → JSON, and its correctness is now particularly load-bearing given the struct was just extended with `HighValue` and `Values`.
> 
> Add at least one content assertion on the filter:
> ```go
> groups := body["filterGroups"].([]interface{})
> f := groups[0].(map[string]interface{})["filters"].([]interface{})[0].(map[string]interface{})
> assert.Equal(t, "hs_task_status", f["propertyName"])
> assert.Equal(t, "EQ", f["operator"])
> assert.Equal(t, "NOT_STARTED", f["value"])
> ```
> The second sub-test already does this depth of inspection for BETWEEN/IN, so the pattern is established.

</details>

<details>
<summary><strong>policies:conventions (1 finding)</strong></summary>

### Minor - `internal/cmd/tasks/tasks.go:363`

> The `newSearchCmd` implementations in `tasks.go` and `emails.go` are nearly identical — they differ only in object type, table headers, and example text. The same PR that introduced `internal/cmd/shared` to hold reusable filter/sort logic did not extract the shared command scaffold, so any future object type (calls, contacts, etc.) would require a third copy.
> 
> Suggested fix: add a constructor in `internal/cmd/shared` (e.g. `NewSearchCmd(opts, objectType, headers, defaultProperties, examples)`) and delegate from each object-specific `newSearchCmd`. This keeps the reuse boundary consistent with what was already established by `ParseFilters`/`ParseSort`.

</details>

*0 PR discussion threads considered. 0 summarized; 0 resolved.*

---
<details>
<summary>Completed in 3m 59s | ~$0.52 (est.) | claude-sonnet-4-6 | cr 0.4.161</summary>

| Field | Value |
|---|---|
| Model | claude-sonnet-4-6 |
| Reviewers | documentation:docs, go:implementation-tests, policies:conventions |
| Engine | claude_cli · claude-sonnet-4-6 |
| Reviewed by | cr · piekstra-dev |
| Duration | 3m 59s wall · 3m 57s compute |
| Cost | ~$0.52 (est.) |
| Tokens | 20 in / 12.5k out |

**Per-workstream usage**

| Workstream | Model | In | Out | Cache read | Cache create | Cost | Duration |
|---|---|---:|---:|---:|---:|---:|---:|
| orchestrator-selection | claude-sonnet-4-6 | 4 | 1.6k | 7.2k | 12.4k | ~$0.07 (est.) | 29s |
| documentation:docs | claude-sonnet-4-6 | 4 | 1.6k | 7.2k | 12.2k | ~$0.07 (est.) | 31s |
| go:implementation-tests | claude-sonnet-4-6 | 4 | 6.9k | 7.2k | 25.9k | ~$0.20 (est.) | 1m 59s |
| policies:conventions | claude-sonnet-4-6 | 4 | 2.1k | 7.2k | 24.2k | ~$0.12 (est.) | 45s |
| orchestrator-rollup | claude-sonnet-4-6 | 4 | 393 | 17.1k | 11.3k | ~$0.05 (est.) | 12s |

</details>

**Review by piekstra-dev — APPROVED** [review]
Automated PR review completed with outcome: approved.

File: api/crm_test.go (hunk 1 above)
Comment (piekstra) [reply api/crm_test.go]: Done in b71b099. Added content assertions to the "search tasks sends filter, sort, and pagination" sub-test, drilling into `filterGroups[0].filters[0]` and checking `propertyName=hs_task_status`, `operator=EQ`, `value=NOT_STARTED` (matching the depth of the BETWEEN/IN sub-test). Also asserted the serialized sort content (`propertyName=hs_timestamp`, `direction=ASCENDING`) while there, so a dropped/empty filter or sort no longer slips through. Verified the strengthened test passes under `go test -race ./...`.

File: internal/cmd/tasks/tasks.go (hunk 2 above)
Comment (piekstra) [reply internal/cmd/tasks/tasks.go:363]: Done in b71b099. Extracted `shared.NewSearchCmd(opts, shared.SearchCmdConfig{...})` in `internal/cmd/shared/searchcmd.go`; both `tasks.go` and `emails.go` now delegate to it, keeping the reuse boundary consistent with `ParseFilters`/`ParseSort`. The object-specific pieces are passed via the config (object type, noun, short/long/example text, default properties, headers, and a `Row func(api.CRMObject) []string` row builder), so a future object type only adds a small config rather than a third copy of the scaffold. The shared constructor owns all the command wiring (flags, filter/sort parsing, request building, pagination, rendering) — no behavior change, and the duplication is gone (net ~130 lines removed). `make build`/`make test`/`make lint`/`go vet` all green.

Comment (piekstra-dev) [reply internal/cmd/tasks/tasks.go:363]: Reviewer noted that newSearchCmd in tasks.go and emails.go were nearly identical, with duplication that would compound for future object types, despite the PR already establishing a shared package for filter/sort logic. Author confirmed fix in b71b099: extracted shared.NewSearchCmd with a SearchCmdConfig struct in internal/cmd/shared/searchcmd.go; both tasks.go and emails.go now delegate to it, removing ~130 lines of duplication while keeping all behavior unchanged.

File: api/crm_test.go (hunk 1 above)
Comment (piekstra-dev) [reply api/crm_test.go]: Reviewer flagged that the 'search tasks sends filter, sort, and pagination' sub-test only checked for the presence of filterGroups/sorts keys without asserting their content, leaving serialisation bugs undetectable. Author confirmed fix in b71b099: added content assertions drilling into filterGroups[0].filters[0] (propertyName=hs_task_status, operator=EQ, value=NOT_STARTED) and sort content (propertyName=hs_timestamp, direction=ASCENDING), matching the depth of the BETWEEN/IN sub-test.

File: internal/cmd/shared/searchcmd.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	cmd.Flags().StringArrayVar(&filterArgs, "filter", nil, "Filter condition (e.g. prop=value, prop>=value, prop:OPERATOR:value); repeatable")
+	cmd.Flags().StringArrayVar(&sortArgs, "sort", nil, "Sort condition (e.g. hs_timestamp:asc or hs_timestamp:desc); repeatable")
+	cmd.Flags().IntVar(&limit, "limit", 10, "Maximum number of results")
+	cmd.Flags().StringVar(&after, "after", "", "Pagination cursor for the next page")
+	cmd.Flags().StringSliceVar(&properties, "properties", nil, "Properties to include (comma-separated)")
+
+	return cmd
+}
```
Comment (piekstra-dev) [inline internal/cmd/shared/searchcmd.go:118]: `--properties` uses `StringSliceVar` while the sibling flags `--filter` and `--sort` in the same command use `StringArrayVar`. `StringSliceVar` splits on commas, so `--properties id,name` expands to two values; `StringArrayVar` would not. This split behaviour is likely intentional for property lists, but the inconsistency within a single command's flag definitions is surprising for contributors and differs from the established pattern used elsewhere in this file. If comma-split is desired, add a note in the flag description or a comment explaining the deliberate choice; otherwise align to `StringArrayVar` and let users repeat the flag (`--properties id --properties name`).

*Reply inline to this comment.*

File: api/crm_test.go [hunk 4]
```diff
@@ -428,6 +429,76 @@ func TestClient_SearchObjects(t *testing.T) {
 
 		require.NoError(t, err)
 	})
+
+	t.Run("search tasks sends filter, sort, and pagination", func(t *testing.T) {
+		var body map[string]interface{}
+		server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			assert.Equal(t, "/crm/v3/objects/tasks/search", r.URL.Path)
+			assert.Equal(t, http.MethodPost, r.Method)
+			require.NoError(t, json.NewDecoder(r.Body).Decode(&body))
```
Comment (piekstra-dev) [inline api/crm_test.go]: `require.NoError(t, json.NewDecoder(r.Body).Decode(&body))` is called inside the `http.HandlerFunc` goroutine, which is not the goroutine running the test function. Go's testing docs explicitly state that `t.FailNow()` (which `require` calls internally) must only be called from the test goroutine; calling it from another goroutine causes a panic or undefined behavior. Replace both occurrences (one in each new sub-test's handler) with `assert.NoError` so the handler merely marks the test failed and lets it drain cleanly:

```go
assert.NoError(t, json.NewDecoder(r.Body).Decode(&body))
```

*Reply inline to this comment.*

File: internal/cmd/shared/search.go
Comment (piekstra-dev) [inline internal/cmd/shared/search.go:1]: `ParseFilters` and `ParseSort` are non-trivial shared utilities with multiple parsing forms (`=`, `!=`, `>=`, `<=`, `>`, `<`, `:OP`, `:OP:val`, `:BETWEEN:lo:hi`, `:IN:a,b,c`), a heuristic operator-token detector, and silent date auto-conversion. No test file appears in this diff. Because every future object type's `search` subcommand will call these functions, bugs in edge cases (e.g., property names containing `=`, URLs as values, the `isOperatorToken` heuristic misfiring) compound across callers without a test safety net. **Fix:** add table-driven unit tests for `ParseFilters` and `ParseSort` covering the documented forms, invalid inputs, and the operator-detection boundary cases before this utility spreads further.

*Reply inline to this comment.*

File: README.md [hunk 5]
```diff
@@ -186,6 +186,27 @@ hspt tasks create --subject "Follow up" --body "Call about renewal" --priority H
...
+
+# Overdue tasks (not started, due on or before a date — ISO dates are converted automatically)
+hspt tasks search --filter "hs_task_status=NOT_STARTED" --filter "hs_timestamp<=2026-03-17" --sort "hs_timestamp:asc"
+
+# Outbound emails, newest first
+hspt emails search --filter "hs_email_direction=EMAIL" --sort "hs_timestamp:desc" --limit 20
+
+# Emails whose subject contains a phrase
+hspt emails search --filter "hs_email_subject:CONTAINS_TOKEN:Dev Academy" --limit 10
+```
```
Comment (piekstra-dev) [inline README.md:204]: The comment says "Outbound emails" but the filter is `hs_email_direction=EMAIL`. In HubSpot's engagement model, `EMAIL` is the engagement *type* (i.e., it is an email engagement record), not the *direction*. The direction property typically holds values like `OUTBOUND` or `INBOUND`. If the intent is truly to filter for outbound emails, the filter should likely be `hs_email_direction=OUTBOUND`; if the intent is just to list email-type engagements, the comment should drop the word "Outbound" to avoid teaching the wrong pattern to readers who copy-paste this example.

*Reply inline to this comment.*

File: internal/cmd/shared/searchcmd.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+			req := api.SearchRequest{
+				Properties: properties,
+				Limit:      limit,
+				After:      after,
+				Sorts:      sorts,
+			}
+			if len(filters) > 0 {
+				req.FilterGroups = []api.SearchFilterGroup{
+					{Filters: filters},
+				}
```
Comment (piekstra-dev) [inline internal/cmd/shared/searchcmd.go:80]: The `RunE` lambda contains non-trivial logic that is not covered by any test in the diff:

1. **Filter grouping** (lines ~82-84): all parsed filters are unconditionally wrapped into a single `SearchFilterGroup`. If `ParseFilters` returns filters but `len(filters) == 0` is false, this wrapping is the only place that contracts multiple `--filter` flags into the correct HubSpot request shape. A future change that accidentally emits multiple groups or drops the wrapping would go undetected.
2. **Empty-result branch** (line ~88): the `v.Info("No %ss found...")` path is never exercised.
3. **Pagination footer** (lines ~107-109): the `"More results available"` message is conditional on `result.Paging != nil && result.Paging.Next != nil` and is likewise untested.

Add a table-driven test for `NewSearchCmd` using a fake/stub `api.Client` (or the existing httptest pattern from `api/crm_test.go`) that covers at least: (a) filters collapsed into one group in the outbound request, (b) empty-result output, and (c) pagination footer when `Paging.Next` is set.

*Reply inline to this comment.*

File: api/crm_test.go [hunk 7]
```diff
@@ -428,6 +429,76 @@ func TestClient_SearchObjects(t *testing.T) {
...
+		})
+
+		require.NoError(t, err)
+		assert.Len(t, result.Results, 1)
+		assert.Equal(t, float64(50), body["limit"])
+		assert.Equal(t, "cursor123", body["after"])
+
+		groups := body["filterGroups"].([]interface{})
+		f := groups[0].(map[string]interface{})["filters"].([]interface{})[0].(map[string]interface{})
+		assert.Equal(t, "hs_task_status", f["propertyName"])
```
Comment (piekstra-dev) [inline api/crm_test.go:463]: The chained bare type-assertions — e.g. `body["filterGroups"].([]interface{})`, `groups[0].(map[string]interface{})["filters"].([]interface{})[0].(map[string]interface{})` — will panic with an unhelpful runtime message rather than a test failure if the decoded shape is unexpected. Use comma-ok form with a `require.True` check, or introduce typed structs for decoding the request body, so a shape mismatch produces a readable test failure instead of a panic.

*Reply inline to this comment.*

File: internal/cmd/shared/search.go [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+	{"<=", "LTE"},
+	{"!=", "NEQ"},
+	{"=", "EQ"},
+	{">", "GT"},
+	{"<", "LT"},
+}
+
+// dateProperties are HubSpot properties whose values are stored as Unix
+// millisecond timestamps. ISO-8601 dates supplied for these properties are
+// converted automatically so users can write human-readable dates.
```
Comment (piekstra-dev) [inline internal/cmd/shared/search.go:28]: The `dateProperties` allowlist is a hardcoded contract baked into a shared parsing utility. When a user supplies a human-readable date for a property that is not in this map (e.g., a custom date property, or a future HubSpot property like `hs_meeting_start_time`), `convertDateValue` silently passes the string through unchanged. The HubSpot API then receives a non-numeric value where it expects Unix milliseconds, producing an API error or—worse—a silent mismatch with no user-visible warning. This list will drift as the API evolves and as new object types are added.

The list also belongs at the wrong layer: date-property semantics are an API contract, not a CLI parsing detail. **Fix:** move the date-property set into the `api` package (or a per-object-type metadata struct on `SearchCmdConfig`) so it is co-located with the types it describes and can be evolved alongside them. Alternatively, suppress silent pass-through: if a value looks like an ISO date but the property is not in the known list, emit a warning rather than silently forwarding the unconverted string.

*Reply inline to this comment.*

**Comment by piekstra-dev** [comment]
## Automated PR Review

**Reviewed commit:** `b71b099fd543`
**Profile:** `reviewer` - **Posting as:** `piekstra-dev`

### Summary

| Reviewer | Findings |
|----------|----------|
| documentation:docs | 1 |
| go:implementation-tests | 2 |
| policies:conventions | 0 |
| structure:repo-health | 2 |

<details>
<summary><strong>documentation:docs (1 finding)</strong></summary>

### Minor - `README.md:204`

> The comment says "Outbound emails" but the filter is `hs_email_direction=EMAIL`. In HubSpot's engagement model, `EMAIL` is the engagement *type* (i.e., it is an email engagement record), not the *direction*. The direction property typically holds values like `OUTBOUND` or `INBOUND`. If the intent is truly to filter for outbound emails, the filter should likely be `hs_email_direction=OUTBOUND`; if the intent is just to list email-type engagements, the comment should drop the word "Outbound" to avoid teaching the wrong pattern to readers who copy-paste this example.

</details>

<details>
<summary><strong>go:implementation-tests (2 findings)</strong></summary>

### Minor - `api/crm_test.go:438`

> `require.NoError(t, json.NewDecoder(r.Body).Decode(&body))` is called inside the `http.HandlerFunc` goroutine, which is not the goroutine running the test function. Go's testing docs explicitly state that `t.FailNow()` (which `require` calls internally) must only be called from the test goroutine; calling it from another goroutine causes a panic or undefined behavior. Replace both occurrences (one in each new sub-test's handler) with `assert.NoError` so the handler merely marks the test failed and lets it drain cleanly:
> 
> ```go
> assert.NoError(t, json.NewDecoder(r.Body).Decode(&body))
> ```

### Minor - `internal/cmd/shared/searchcmd.go:80`

> The `RunE` lambda contains non-trivial logic that is not covered by any test in the diff:
> 
> 1. **Filter grouping** (lines ~82-84): all parsed filters are unconditionally wrapped into a single `SearchFilterGroup`. If `ParseFilters` returns filters but `len(filters) == 0` is false, this wrapping is the only place that contracts multiple `--filter` flags into the correct HubSpot request shape. A future change that accidentally emits multiple groups or drops the wrapping would go undetected.
> 2. **Empty-result branch** (line ~88): the `v.Info("No %ss found...")` path is never exercised.
> 3. **Pagination footer** (lines ~107-109): the `"More results available"` message is conditional on `result.Paging != nil && result.Paging.Next != nil` and is likewise untested.
> 
> Add a table-driven test for `NewSearchCmd` using a fake/stub `api.Client` (or the existing httptest pattern from `api/crm_test.go`) that covers at least: (a) filters collapsed into one group in the outbound request, (b) empty-result output, and (c) pagination footer when `Paging.Next` is set.

</details>

<details>
<summary><strong>structure:repo-health (2 findings)</strong></summary>

### Major - `internal/cmd/shared/search.go:28`

> The `dateProperties` allowlist is a hardcoded contract baked into a shared parsing utility. When a user supplies a human-readable date for a property that is not in this map (e.g., a custom date property, or a future HubSpot property like `hs_meeting_start_time`), `convertDateValue` silently passes the string through unchanged. The HubSpot API then receives a non-numeric value where it expects Unix milliseconds, producing an API error or—worse—a silent mismatch with no user-visible warning. This list will drift as the API evolves and as new object types are added.
> 
> The list also belongs at the wrong layer: date-property semantics are an API contract, not a CLI parsing detail. **Fix:** move the date-property set into the `api` package (or a per-object-type metadata struct on `SearchCmdConfig`) so it is co-located with the types it describes and can be evolved alongside them. Alternatively, suppress silent pass-through: if a value looks like an ISO date but the property is not in the known list, emit a warning rather than silently forwarding the unconverted string.

### Minor - `internal/cmd/shared/search.go`

> `ParseFilters` and `ParseSort` are non-trivial shared utilities with multiple parsing forms (`=`, `!=`, `>=`, `<=`, `>`, `<`, `:OP`, `:OP:val`, `:BETWEEN:lo:hi`, `:IN:a,b,c`), a heuristic operator-token detector, and silent date auto-conversion. No test file appears in this diff. Because every future object type's `search` subcommand will call these functions, bugs in edge cases (e.g., property names containing `=`, URLs as values, the `isOperatorToken` heuristic misfiring) compound across callers without a test safety net. **Fix:** add table-driven unit tests for `ParseFilters` and `ParseSort` covering the documented forms, invalid inputs, and the operator-detection boundary cases before this utility spreads further.

</details>

*2 PR discussion threads considered. 2 summarized; 2 resolved.*

---
<details>
<summary>Completed in 4m 47s | ~$0.66 (est.) | claude-sonnet-4-6 | cr 0.4.161</summary>

| Field | Value |
|---|---|
| Model | claude-sonnet-4-6 |
| Reviewers | documentation:docs, go:implementation-tests, policies:conventions, structure:repo-health |
| Engine | claude_cli · claude-sonnet-4-6 |
| Reviewed by | cr · piekstra-dev |
| Duration | 4m 47s wall · 4m 45s compute |
| Cost | ~$0.66 (est.) |
| Tokens | 24 in / 13.3k out |

**Per-workstream usage**

| Workstream | Model | In | Out | Cache read | Cache create | Cost | Duration |
|---|---|---:|---:|---:|---:|---:|---:|
| orchestrator-selection | claude-sonnet-4-6 | 4 | 2.3k | 7.2k | 15.0k | ~$0.09 (est.) | 45s |
| documentation:docs | claude-sonnet-4-6 | 4 | 1.5k | 7.2k | 12.2k | ~$0.07 (est.) | 28s |
| go:implementation-tests | claude-sonnet-4-6 | 4 | 3.5k | 7.2k | 25.9k | ~$0.15 (est.) | 1m 13s |
| policies:conventions | claude-sonnet-4-6 | 4 | 3.3k | 7.2k | 23.4k | ~$0.14 (est.) | 1m 04s |
| structure:repo-health | claude-sonnet-4-6 | 4 | 2.2k | 7.2k | 19.3k | ~$0.11 (est.) | 45s |
| orchestrator-rollup | claude-sonnet-4-6 | 4 | 548 | 8.3k | 22.6k | ~$0.10 (est.) | 28s |

</details>

**Review by piekstra-dev — COMMENTED** [review]
Automated PR review completed with outcome: comment.

File: api/crm_test.go (hunk 4 above)
Comment (piekstra) [reply api/crm_test.go]: Fixed in fbe75f7. Both in-handler `require.NoError(t, json.NewDecoder(r.Body).Decode(&body))` calls (one in each new sub-test) are now `assert.NoError`, so the handler only marks the test failed and drains rather than calling `t.FailNow()` off the test goroutine. The `require.NoError(t, err)` assertions in the test goroutine itself are left unchanged.

File: internal/cmd/shared/search.go
Comment (piekstra) [reply internal/cmd/shared/search.go:1]: Tests already exist. `internal/cmd/shared/search_test.go` (added earlier in this PR, b71b099) has table-driven tests for both `ParseFilters` and `ParseSort` covering all documented forms (`=`, `!=`, `>=`, `<=`, `>`, `<`, explicit `:OP`, `:OP:val`, `:BETWEEN:lo:hi`, `:IN`/`:NOT_IN`), the `isOperatorToken` boundary (lowercase token treated as value, missing property name), ISO date auto-conversion on date vs non-date properties, and invalid-input error cases. The re-review only saw the polish diff, so the test file was not in view. Resolving.

File: README.md (hunk 5 above)
Comment (piekstra) [reply README.md:204]: Fixed in fbe75f7 by dropping "Outbound" from the comment. `hs_email_direction` is an enum whose values are `EMAIL` (a logged/sent email) and `INCOMING_EMAIL` — not free-form `OUTBOUND`/`INBOUND` — so `=EMAIL` is a valid direction filter but does not cleanly map to "outbound" semantics across configs. The example now reads "Logged emails by direction, newest first", which is accurate and avoids teaching a misleading pattern. Resolving.

File: internal/cmd/shared/searchcmd.go (hunk 3 above)
Comment (piekstra) [reply internal/cmd/shared/searchcmd.go:118]: Deliberate, leaving as-is. The comma-split from `StringSliceVar` is intentional for `--properties` so users can write `--properties id,name,hs_task_subject` the way HubSpot property lists are normally expressed (and the way the comma-separated `properties` query param is documented). `--filter`/`--sort` use `StringArrayVar` precisely because their values can legitimately contain commas (e.g. `:IN:a,b,c`), where comma-splitting would be wrong. So the inconsistency reflects a real semantic difference rather than an oversight. The flag help text already says "(comma-separated)". Noting as deferred/non-blocking on this approved PR; resolving.

File: internal/cmd/shared/searchcmd.go (hunk 6 above)
Comment (piekstra) [reply internal/cmd/shared/searchcmd.go:80]: Noted as an optional follow-up; non-blocking. A clean test of the `RunE` lambda (filter grouping into a single group, empty-result `v.Info`, and the pagination footer) currently needs a seam to inject a stub API client: `opts.APIClient()` returns a concrete `*api.Client` built from global config, so there is no interface to stub without either a config/env + httptest setup or a small refactor to make the client injectable. That is a larger change than is warranted on an already-approved PR, so deferring it as a future follow-up rather than implementing it here. The underlying request-shape behaviour (filters collapsed into one group, BETWEEN/IN serialization) is covered at the API layer in `api/crm_test.go`. Resolving.

File: api/crm_test.go (hunk 7 above)
Comment (piekstra) [reply api/crm_test.go:463]: Noted as an optional test-robustness follow-up; non-blocking. The chained type-assertions on the decoded request body would panic (rather than fail readably) only if the request shape were unexpected, but in this test the body is produced by our own `SearchObjects` serialization, so a shape mismatch would itself be the bug under test and a panic still fails the run. Converting these to comma-ok `require.True` checks or a typed decode struct is a reasonable readability improvement; deferring it as a follow-up rather than expanding this final pass. Resolving.

File: internal/cmd/shared/search.go (hunk 8 above)
Comment (piekstra) [reply internal/cmd/shared/search.go:28]: Acknowledged as a reasonable future improvement, but out of scope for this PR. Agreed that date-property semantics are really an API-layer contract; moving the `dateProperties` set into the `api` package (or onto per-object-type metadata, e.g. `SearchCmdConfig`) so it evolves alongside the types it describes is the right long-term shape. That is a larger design refactor than fits this final, approved polish pass. The current behaviour is a safe superset: unknown properties pass the value through unchanged, matching the pre-existing default, so this is additive rather than a regression. Capturing as a future follow-up; resolving.
